PSP for JSON-RPC

lamafab opened the pull request:
(UPDATED 2020-11-25) Preview: https://github.com/w3f/PSPs/blob/psp-rpc-api/PSPs/drafts/psp-6.md

As discussed, this is a draft/WIP and will change a lot over the next days/weeks. Generally, it needs to adjust to the decisions made by Parity.

~~A TODO list available here: https://github.com/w3f/polkadot-spec/issues/67#issuecomment-597658477~~

FlorianFranzen commented:
Quick update: Soramitsu [indicated](
https://github.com/soramitsu/kagome/issues/401#issuecomment-678939877) that there are a few essential calls missing, namely `rpc_methods`, `state_getMetadata` and `system_chainType`.

Noc2 reviewed (changes requested):
Btw. the psp number should be 6 (same as PR) and feel free to change the status to draft. After this, I think we can merge it into the draft folder: https://github.com/w3f/PSPs/tree/master/PSPs/drafts This way it's easier to discuss it and others can contribute to it.

FlorianFranzen commented on PSPs/drafts/psp-6.md:
````diff
@@ -0,0 +1,2903 @@
+Request:
+
+```bash
+curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "chain_getBlockHash", "params":[50, "0x64", [200]]}' http://localhost:9933
````
One of seem to suggest just one of them, here the example uses all three at once. Is that as intended?

lamafab replied:
The parameters on this method work somewhat interestingly. You can specify one or multiple block numbers, either as integers or hex string, and can even include embedded arrays. But I guess this is not really "good design", so I'll adjust this (backwards compatible with substrate).

FlorianFranzen reviewed (changes requested):
If you fix the last few typos and add a proper section on websockets and pubsub, I think this should be ready to be merged.

danforbes commented on PSPs/drafts/psp-6.md:
```diff
@@ -0,0 +1,2903 @@
+
+A string which represents a 32-byte, hexadecimal-encoded Blake2 hash. Contains a `0x` prefix to indicate the hexadecimal encoding.
+
+#### 1.3.3.7. HEX
+
+A string of varying size representing hexadecimal-encoded data. Contains a `0x` prefix to indicate the hexadecimal encoding.
```
Would it be better to put this before `HASH` since all hashes are hex? Also, consider explicitly enumerating the valid characters, e.g. `[a-fA-F0-9]`

```suggestion
#### 1.3.3.7. HEX

A variable-size string that represents hexadecimal-encoded data. Must begin with the characters `0x` to indicate a hexadecimal encoding, e.g. `0x1337`.
```

lamafab replied:
> Would it be better to put this before HASH since all hashes are hex?

Done!

> Also, consider explicitly enumerating the valid characters, e.g. [a-fA-F0-9]

Given that it explicitly states "hexadecimal-encoded", I guess this is somewhat redundant. But I added it anyway.

danforbes commented on PSPs/drafts/psp-6.md:
```diff
@@ -0,0 +1,2903 @@
+
+RPC error codes are in the `2000` - `2999` range. No known error types are specified.
+
+### 1.5.2. system_name
```
For names like this consider formatting as code...

```suggestion
### 1.5.2. `system_name`
```

danforbes commented on PSPs/drafts/psp-6.md:
```diff
@@ -0,0 +1,2903 @@
+
+## 1.5. System
+
+System RPC API.
```
Can this description be improved a bit? It doesn't feel like it adds much at this time. Maybe something along the lines of this?
```suggestion
RPC endpoints that expose system metadata or can be used for administration of the blockchain network.
```

danforbes commented on PSPs/drafts/psp-6.md:
```diff
@@ -0,0 +1,2903 @@
+
+### 1.5.3. system_version
+
+Get the node implementation's version. Should be a semver string.
```
Watch out for jargon like "semver"...if you use it, link out to a resource where the reader can learn more about it.

danforbes reviewed (commented):
Amazing job, @lamafab :bow: Very well-written :tada: I did not perform an exhaustive review; just left a few nitpicky comments. I would also suggest using a formatting tool like Prettier on this in order to limit the line length...this should be enforced on this repo in general, I think, because otherwise documents can be hard to review.

lamafab commented on PSPs/drafts/psp-6.md:
```diff
@@ -0,0 +1,2903 @@
+
+All parameters in the specified JSON-RPC methods are **REQUIRED**, unless explicitly mentioned otherwise (**OPTIONAL**). Return values which are OPTIONAL indicate the possibility of `null` being returned.
+
+Additionally, the JSON-RPC API should be made available over the [WebSocket Protocol](https://tools.ietf.org/html/rfc6455).
```
Done. Just note that there is no official websocket `pubsub` standard.

lamafab commented on PSPs/drafts/psp-6.md:
```diff
@@ -0,0 +1,2903 @@
+#### 1.11.5.2. Response
+
+* `ARRAY`
+  * `STRING` - (OPTIONAL) Storage key
```
Done, also found other sections where this was the case.

lamafab commented on PSPs/drafts/psp-6.md:
````diff
@@ -0,0 +1,2903 @@
+    "method": "grandpa_subscribeJustifications",
+    "params": []
+}
+```
````
I'm not sure about this. Curl examples are quite useful since people can just copy paste those in their own environment (which runs Substrate/Polkadot, for example). With websockets it's not that easy. Personally I used [wscat](https://github.com/websockets/wscat).

Looks like there is a `-x` parameter which can be used as a single command. I can adjust the examples if you think this should be added, because you're right about readability.

lamafab commented on PSPs/drafts/psp-6.md:
```diff
@@ -0,0 +1,2903 @@
+#### 1.8.8.1. Parameter
+
+* `HEX` - The public key.
+* `STRING` * [Key type](#Key-types).
```
I have `*` pretty much everywhere for lists (primarily because my markdown plugin was complaining about `-`). Should I adjust this document-wide?

lamafab replied:
Well, `prettier` took care of that part.

lamafab commented on PSPs/drafts/psp-6.md:
```diff
@@ -0,0 +1,2903 @@
+
+#### 1.3.3.4. MAP
+
+A key/value object.
```
Yes, I added this section:

> A key/value object, where keys are `STRING`s and values are arbitrary types. Each API defines its own value types.

lamafab commented:
@FlorianFranzen @danforbes Thank you for your feedback, I adjusted pretty much everything. Do note that I added a simple `prettier` config (maybe we can just "stage" this via Github Actions and have it run `prettier` on PRs rather than making this part of the repo?).

You can run it with: `yarn prettier --write .`

Also, my editor currently generated the table of contents, including section numbering. Maybe that's something we can do globally for all documents by using `prettier`. I'll look into this some more.

@FlorianFranzen I will update `payment_queryInfo` tomorrow morning. There are currently no docs on this.
